### ISSUE TITLE
Divergence of (hu, hv, h)

### ISSUE THREAD

**Issue by ningliu-iga** [SpeedyWeather/SpeedyWeather.jl#372]
Hi, awesome work here! Thanks for all the development efforts!

I'm trying to use SpeedyWeather shallow water model to generate data for a ML task. The data I generated is in the form of (u, v, h), where h is the pres output added by the layer thickness H minus the orography Hb. I tried to check if the generated data (uh, vh, h) is divergence free by computing div = cos(theta)*dh/dt + d(uh)/d(lambda) + d(vh*cos(theta))/d(theta). The ranges of longitude lambda and latitude theta are [0, 2*pi) and [1.5458759662006802, -1.5458759662006802], respectively, where the range of theta is obtained from the locations of the gauss legendre points in your code. I got from your code that dt=0.001032805 starting from the 2nd time step. In order to maintain a constant dt, the data I used is started from the 5th step. I used fft to compute derivatives with the domain padded by fourier continuation to make it periodic. However, my computed divergence is far from zero. 

Could you please help me point out if there is anything wrong in the data generation or the divergence calculation process? Many thanks in advance!

**Comment by milankl** [comment]
Great to hear you're making use of it! 

From what you write I see you've been using the `ShallowWaterModel`, for which continuity equation reads as (no forcing)
```math
\frac{\partial \eta}{\partial t} + \nabla \cdot (\mathbf{u}h) = 0 \qquad \qquad (1)
```
(from documentation: https://speedyweather.github.io/SpeedyWeather.jl/dev/shallowwater/#Shallow-water-equations)

Which is essentially what you wrote down. Although I wouldn't call it divergence free because the divergence is only the second term, not the time derivative. In the shallow water system a converging flow moves up the interface $\partial_t \eta > 0$. You are right that the divergence operator should be applied in spherical coordinates, but you're also missing the radius: https://speedyweather.github.io/SpeedyWeather.jl/dev/spectral_transform/#Derivatives-in-spherical-coordinates which seems however to be compensated because your dt = 0.001 (I don't know which resolution you run at) seems to be the scaled time step (which is divided by the radius)!

Technically you can also assess mass conservation by integrating (1) over the sphere. #323 got really close to closing this budget and some of the discussion in there might also be helpful here.

I wouldn't be surprised if you can only verify the continuity equation if you do it exactly as it's done in the model: Meaning your time derivative should come from two consecutive time steps and the u,v,h should be provided on the grid that is also used inside the model. The default output is on th `FullGaussianGrid` but the default computational grid is the `OctahedralGaussianGrid`, meaning that the output by default is interpolated. Then the model calculates the gradients in spectral space, not in grid-point space. I don't know how far you are off (your `div` should be zero), feel free to throw any calculations in here.

**Comment by milankl** [comment]
I first thought you were looking for a divergence-flow, for which you should use the `BarotropicModel` which is divergence free as only vorticity is a prognostic variable and so $\nabla \cdot \mathbf{u} = 0$.

**Comment by milankl** [comment]
If you want to see some details of how it's done inside the model check out

https://github.com/SpeedyWeather/SpeedyWeather.jl/blob/1564fe95a90e1ea48d8ca62fe0c2cf5911e99cdc/src/dynamics/tendencies_dynamics.jl#L552

which calculates `h`

https://github.com/SpeedyWeather/SpeedyWeather.jl/blob/1564fe95a90e1ea48d8ca62fe0c2cf5911e99cdc/src/dynamics/tendencies_dynamics.jl#L358

and this computes the volume flux divergence $-\nabla \cdot (\mathbf{u}h)$ and adds that to the tendency

**Comment by milankl** [comment]
Also I don't know which version you are using (please use #main, most of the documentation is now written for that, v0.6 isn't released yet, but we're close to it!) but you can always check the time step after model creation with
```julia
julia> spectral_grid = SpectralGrid(trunc=31)
SpectralGrid:
 Spectral:   T31 LowerTriangularMatrix{Complex{Float32}}, radius = 6.371e6 m
 Grid:       3168-element, 48-ring OctahedralGaussianGrid{Float32} (quadratic)
 Resolution: 401km (average)
 Vertical:   8-level SigmaCoordinates

julia> model = ShallowWaterModel(;spectral_grid);

julia> model.time_stepping
Leapfrog{Float32}:
 trunc::Int64 = 31
 Δt_at_T31::Float64 = 30.0
 radius::Float32 = 6.371e6
 robert_filter::Float32 = 0.05
 williams_filter::Float32 = 0.53
 Δt_sec::Int64 = 1800
 Δt::Float32 = 0.0002825302
 Δt_hrs::Float64 = 0.5
```

`Δt` is the scaled time step, `Δt_sec` is the (unscaled) time step in seconds which at the default settings is 30min

**Comment by ningliu-iga** [comment]
Many thanks for your quick response, Milan. 

Yes, you're correct that I am trying to verify the data using the continuity equation. I say divergence free because the data (hu, hv, h) should be divergence free with respect to (lambda the longitude, theta the latitude, t the time), which essentially is the continuity equation. 

I skipped the radius term since your documentation says the continuity equation is scaled by R. I directly obtained dt by printing out 2Δt from the main loop of the time_stepping function. The spatial resolution I use is 192*96, and the temporal one should be 6 hrs if I'm not wrong. I ran SpeedyWeather using run_speedy(Float32; n_days=20, model=:shallowwater, output=true, trunc=62, Δt_at_T85=40, initial_conditions=:random2, out_path=loc). Unfortunately, the version I use is v0.2.1, which is definitely not the most recent version. This version was installed on my machine together with another package (pdearena) which uses SpeedyWeather to generate ML data as well. 

For computing the time derivative, I used FFT to compute the numerical derivative. I already verified my algorithm using analytical solutions so the derivative calculation should be correct. For the (pres, u, v) data, I directly got them from the output, so I assume they are the interpolated data on FullGaussianGrid. I managed to output the H-Hb data by inserting the following code in the time_stepping function right before first_timesteps is called:

```G = M.geospectral
	@unpack nlon, nlat = G.geometry
	@unpack lond,latd = M.geospectral.geometry
	@unpack output_startdate, compression_level = M.parameters
	H₀ = M.constants.layer_thickness
	H₀ = convert(NF,H₀)
	B = M.boundaries
	@unpack orography = B

	h_add = zeros(nlon, nlat) 
	@inbounds for j in 1:nlat
        for i in 1:nlon
            # h = η + H₀ - orography
            h_add[i,j] = H₀ - orography[i,j]
        end
    end ```

Then I do 
```NetCDF.putvar(netcdf_file,"h_add",h_add)
   write_netcdf_output!(netcdf_file,feedback,time_sec,diagn,M)```
to save the h_add data together with the original output.

I'm not a Julia guy, so I wouldn't be surprised if I made any mistake in the code above. Maybe the h_add data is not on the FullGaussianGrid? Please help me confirm.

**Comment by milankl** [comment]
Okay v0.2.1 is waaaay old, and I cannot recall the state of the shallow water model therein. Please update to #main and treat it is v0.6. I remember a discussion with a group that put a paper on arxiv using that version but in the end I believe they only resolved the initial shock and no proper shallow water dynamics. You can install the main branch as outlined in the documentation https://speedyweather.github.io/SpeedyWeather.jl/dev/installation/

v0.2.1 only has the FullGaussianGrid I believe, if you want to keep that in the newer versions you'd need to specify `Grid=FullGaussianGrid` as the current default is the `OctahedralGaussianGrid`. See also documentation on those.

In the `ShallowWaterModel` the thickness H is a property of the atmosphere
```julia
julia> atmosphere = EarthAtmosphere()
EarthAtmosphere:
 mol_mass_dry_air::Float64 = 28.9649
 mol_mass_vapour::Float64 = 18.0153
 cₚ::Float64 = 1004.0
 R_gas::Float64 = 8.3145
 R_dry::Float64 = 287.05433127682124
 R_vapour::Float64 = 461.52437095135804
...
 layer_thickness::Float64 = 8.5
```
given in kilometers. You can always get the orography (after initialize!) like so
```julia
julia> spectral_grid = SpectralGrid(trunc=31,nlev=1,Grid=FullGaussianGrid);
julia> model = ShallowWaterModel(;spectral_grid);
julia> simulation = initialize!(model);
julia> model.orography.orography
4608-element, 48-ring FullGaussianGrid{Float32}:
  -73.18947
  -73.670944
  -73.26136
  -71.965225
    ⋮
 2502.2954
 2544.1804
 2587.757
 2632.8286
```
(values in meters) You can also plot it like so

![image](https://github.com/SpeedyWeather/SpeedyWeather.jl/assets/25530332/b606ffe0-4b5a-45fd-9eb1-4074480673a8)

(or any other gridded data)

From v0.6 there should be no need to hack the functions as you do ;) The entire model state is always available in the terminal / in a notebook and you can access things there directly in an interactive way.

**Comment by ningliu-iga** [comment]
Okay I see. Thanks a lot for the insight! My main purpose of using SpeedyWeather is to generate the dataset (hu, hv, h) that is divergence free with respect to (lambda, theta, t). However, since my ML model handles data in the Fourier space, the grid that the data is on should be uniform. Any suggestion how I can get the (hu, hv, h) data on a uniform grid? The longitude should be already uniform so the issue is the latitude grid. If there is no way to output data on a uniform latitude grid, I can possibly do some postprocessing using high-order spline interpolation. Thanks again!

**Comment by milankl** [comment]
> Okay I see. Thanks a lot for the insight! My main purpose of using SpeedyWeather is to generate the dataset (hu, hv, h) that is divergence free with respect to (lambda, theta, t). In other words, the data should satisfy the continuity equation. 

I still struggle to understand why this is satisfies the general conception of "divergence-free" -- like there's a prognostic variable called divergence and it's not zero. But that seems more a different use of terminology...!

> However, since my ML model handles data in the Fourier space, the grid that the data is on should be uniform. Any suggestion how I can get the (hu, hv, h) data on a uniform grid? The longitude should be already uniform so the issue is the latitude grid. If there is no way to output data on a uniform latitude grid, I can possibly do some postprocessing using high-order spline interpolation. Thanks again!

By uniform you mean evenly spaced in both lat and lon direction? (there's also "uniform" in terms of equal-area, but that doesn't seem to be your concern). We have a `FullClenshawGrid` (see [Grids documentation](https://speedyweather.github.io/SpeedyWeather.jl/dev/grids/), especially https://speedyweather.github.io/SpeedyWeather.jl/dev/grids/#FullClenshawGrid) that you can use for computations and/or output. It's not in v0.2.1 I believe, but in #main you can do

```julia
julia> spectral_grid = SpectralGrid(Grid=FullClenshawGrid)
SpectralGrid:
 Spectral:   T31 LowerTriangularMatrix{Complex{Float32}}, radius = 6.371e6 m
 Grid:       4512-element, 47-ring FullClenshawGrid{Float32} (quadratic)
 Resolution: 336km (average)
 Vertical:   8-level SigmaCoordinates
```

And you'll get all computations and output on a 3.75˚ x 3.75˚ grid (excluding the poles)

**Comment by milankl** [comment]
The resolution parameter for every grid is `nlat_half` which you can use to obtain all sorts of information for that grid and resolution
```julia
julia> nlat_half = spectral_grid.nlat_half
24

julia> RingGrids.get_lond(FullClenshawGrid,nlat_half)
96-element Vector{Float64}:
   0.0
   3.75
   7.5
   ⋮
 348.75
 352.5
 356.25

julia> RingGrids.get_latd(FullClenshawGrid,nlat_half)
47-element Vector{Float64}:
  86.25
  82.5
  78.75
  75.0
  71.25
  67.5
   ⋮
 -67.5
 -71.25
 -75.0
 -78.75
 -82.5
 -86.25
```

**Comment by ningliu-iga** [comment]
Thank you so much, Milan, not just for the awesome software, but also for the strong user support! I will try these options and keep you posted whether or not I can generate data that satisfy the continuity equation.

**Comment by milankl** [comment]
Well I'm sure it does, otherwise the mass conservation from #323 would be way off, but how to verify this easily externally is the question ;)

**Comment by ningliu-iga** [comment]
Oh, just want to double check, does v0.5 contain all the above features too? I followed your installation instruction and the version installed is v0.5. If not, can I just replace the installed source code with the code in #main?

**Comment by ningliu-iga** [comment]
> Well I'm sure it does, otherwise the mass conservation from #323 would be way off, but how to verify this easily externally is the question ;)

Yeah I agree. The data is generated using the continuity equation, so of course it's gonna satisfy that. The challenge is how to properly handle the generated data and compute the derivatives with another numerical approach.

**Comment by milankl** [comment]
The package manager (open with `]` check list of installed packages with `st` or `status`) should say
```julia
[9e226e20] SpeedyWeather v0.5.0 `https://github.com/SpeedyWeather/SpeedyWeather.jl#main`
```
Then you're good to go! It's technically still v0.5, but well be soon tagged as v0.6. The released v0.5 has another interface...

**Comment by ningliu-iga** [comment]
Thanks Milan. This is what I got after updating SpeedyWeather. It doesn't say explicitly that the version corresponds to #main. Is that fine?

```
(@v1.9) pkg> st
Status `~/.julia/environments/v1.9/Project.toml`
  [9e226e20] SpeedyWeather v0.5.0

(@v1.9) pkg> update SpeedyWeather
    Updating registry at `~/.julia/registries/General.toml`
  No Changes to `~/.julia/environments/v1.9/Project.toml`
  No Changes to `~/.julia/environments/v1.9/Manifest.toml`
[ Info: We haven't cleaned this depot up for a bit, running Pkg.gc()...
      Active manifest files: 2 found
      Active artifact files: 23 found
      Active scratchspaces: 0 found
     Deleted no artifacts, repos, packages or scratchspaces

(@v1.9) pkg> st
Status `~/.julia/environments/v1.9/Project.toml`
  [9e226e20] SpeedyWeather v0.5.0

```

**Comment by milankl** [comment]
nope that's the latest release that still has another interface... do
```julia
(@v1.9) pkg> add https://github.com/SpeedyWeather/SpeedyWeather.jl#main
```

**Comment by ningliu-iga** [comment]
Got it thanks! BTW, I sent you an email. Just want to let you know here in case it slipped through your email..

**Comment by milankl** [comment]
@ningliu-iga just released v0.6, you can use that now!

**Comment by ningliu-iga** [comment]
Sounds great. Thanks for letting me know!

**Comment by milankl** [comment]
I want to keep this issue for the divergence verification, the initial conditions are discussed in #377

**Comment by milankl** [comment]
The continuity equation can be closed externally by 0.1% RMSE 

![image](https://github.com/SpeedyWeather/SpeedyWeather.jl/assets/25530332/4c0910cf-8f4c-4d23-a80b-b44f077949e4)

with the following

1. get $h$ from $\eta$
```julia
η = Array{Float32}(ds["pres"][:,:,:]);
H = model.constants.layer_thickness
orography = repeat(Matrix{Float32}(ds["orography"][:]),1,1,ds.dim["time"])
h = η .+ H .- orography
```

2. Centred derivative in time, as the model uses leapfrogging
```julia
Δt = Dates.Second(ds["time"][2] - ds["time"][1]).value
∂η_∂t = (η[:,:,3:end] .- η[:,:,1:end-2])/2Δt;
```
One can switch off the Williams filter, but not the Robert filter (unstable). I didn't find this helpful though for further accuracy.

3. Volume flux divergence via spectral space
```julia
u = Array{Float32}(ds["u"][:,:,1,:])
v = Array{Float32}(ds["v"][:,:,1,:]);
uh = u .* h
vh = v .* h;

k = 1000   # some time step

# at k-1
Grid = spectral_grid.Grid
uh_grid = Grid(uh[:,:,k-1])
RingGrids.scale_coslat⁻¹!(uh_grid)
uh_spec = SpeedyWeather.spectral(uh_grid,model.spectral_transform);

vh_grid = Grid(vh[:,:,k-1])
RingGrids.scale_coslat⁻¹!(vh_grid)
vh_spec = SpeedyWeather.spectral(vh_grid,model.spectral_transform);

div = zeros(LowerTriangularMatrix{Complex{spectral_grid.NF}},65,64)
SpeedyTransforms.divergence!(div,uh_spec,vh_spec,model.spectral_transform)
div_grid = Matrix(SpeedyWeather.gridded(div,model.spectral_transform))
div_grid ./= model.spectral_grid.radius;

divm1 = Matrix(div_grid)

# at k+1
uh_grid = Grid(uh[:,:,k+1])
RingGrids.scale_coslat⁻¹!(uh_grid)
uh_spec = SpeedyWeather.spectral(uh_grid,model.spectral_transform);

vh_grid = Grid(vh[:,:,k+1])
RingGrids.scale_coslat⁻¹!(vh_grid)
vh_spec = SpeedyWeather.spectral(vh_grid,model.spectral_transform);

div = zeros(LowerTriangularMatrix{Complex{spectral_grid.NF}},65,64)
SpeedyTransforms.divergence!(div,uh_spec,vh_spec,model.spectral_transform)
div_grid = Matrix(SpeedyWeather.gridded(div,model.spectral_transform))
div_grid ./= model.spectral_grid.radius;

divm2 = Matrix(div_grid)
divm = 1/2*divm1 .+ 1/2*divm2
```
Import here is to do this twice at $k-1$ and $k+1$ and average, as this corresponds to the semi-implicit algorithm with $\alpha=0.5$

![image](https://github.com/SpeedyWeather/SpeedyWeather.jl/assets/25530332/ce72a47d-f759-49fa-9ff7-99676f86bc06)

4. RMSE of relative error wrt speed with coslat-weighting
```julia
speed = sqrt.(u[:,:,k].^2 + v[:,:,k].^2)
error = abs.(∂η_∂t[:,:,k-1] + divm) ./ speed

error_average = copy(error)
RingGrids.scale_coslat!(FullClenshawGrid(error_average))
RMSE = sqrt.(sum(error_average.^2)/length(error))
```

Model setup for data generation (use current main branch)
```julia
using SpeedyWeather
spectral_grid = SpectralGrid(trunc=63,nlev=1,Grid=FullClenshawGrid)
time_stepping = SpeedyWeather.Leapfrog(spectral_grid,Δt_at_T31=15)
implicit = SpeedyWeather.ImplicitShallowWater(spectral_grid,α=0.5)
orography = EarthOrography(spectral_grid,smoothing=false)
initial_conditions = SpeedyWeather.RandomWaves()
output = OutputWriter(spectral_grid,ShallowWater,output_dt=1/8,output_vars=[:u,:v,:pres,:orography])

model = ShallowWaterModel(;spectral_grid,orography,output,initial_conditions,implicit,time_stepping);
simulation = initialize!(model);
run!(simulation,n_days=10,output=true)
```
